Add keypoints and segments support to Albumentations augmentations

onuralpszr opened the pull request:
Fix: https://github.com/ultralytics/ultralytics/issues/22797

<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

✨ Expanded Albumentations support across all YOLO tasks and added custom Albumentations transforms for classification.

### 📊 Key Changes

- 🧩 Enabled Albumentations augmentations for detection, segmentation, classification, pose, and OBB tasks.
- 🎯 Added synchronized transformation of bounding boxes, segmentation points, and pose keypoints.
- 🔗 Preserved instance alignment when augmentations filter or remove bounding boxes.
- 👁️ Updated pose keypoint visibility when points are warped outside the image.
- 🖼️ Added an `augmentations` option to `classify_augmentations()` for custom Albumentations pipelines.
- 🛡️ Added graceful warnings when Albumentations is unavailable or custom transforms fail to initialize.
- 📚 Updated documentation to reflect broader task compatibility.

### 🎯 Purpose & Impact

- ✅ Provides more consistent and accurate augmentations for multi-instance tasks, reducing mismatches between images and labels.
- 🚀 Gives classification users greater control over training augmentation through custom Albumentations transforms.
- 🧠 Improves support for advanced YOLO workflows, including pose and segmentation training.
- ⚠️ Existing users should see improved label handling, while classification users can opt into custom augmentations by supplying an `augmentations` configuration.

onuralpszr committed:
feat: 🚀 Enhance data augmentation support for YOLO tasks with keypoints and segments

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge 82c0a6fa3f36bbc7a4a6c3419fc773f5d760e364 into 4ac2d6152afbdf1d0c39649e6f6d4f7e299f1fa3

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant commented on ultralytics/data/augment.py:
```diff
@@ -18,7 +18,7 @@
 from ultralytics.utils.instance import Instances
 from ultralytics.utils.metrics import bbox_ioa
-from ultralytics.utils.ops import segment2box, xywh2xyxy, xyxyxyxy2xywhr
+from ultralytics.utils.ops import segment2box, xywh2xyxy, xyxyxyxy2xywhr, masks2segments
```
❗ **CRITICAL**: `polygons2masks` is used later when converting segments to masks, but it is not imported. This will raise a `NameError` as soon as segment data is processed.

**Suggested change:**
```suggestion
from ultralytics.utils.ops import masks2segments, polygons2masks, segment2box, xywh2xyxy, xyxyxyxy2xywhr
```

UltralyticsAssistant commented on ultralytics/data/augment.py:
```diff
@@ -2011,12 +2015,62 @@ def __call__(self, labels: dict[str, Any]) -> dict[str, Any]:
+                    # Update keypoints if they were transformed
+                    if keypoints is not None and len(keypoints) and "keypoints" in new:
+                        # Reshape back to (N, num_kpts, 2) and restore visibility if it existed
+                        num_instances = len(cls)
```
❗ **CRITICAL**: After Albumentations runs, some bboxes may be dropped. Using `len(cls)` here can then mismatch the length of `new["keypoints"]`, causing a reshape failure. Please base the reshape on the transformed annotations (`new["class_labels"]` or `new["bboxes"]`) so the array sizes stay aligned.

**Suggested change:**
```suggestion
                        num_instances = len(new["class_labels"])
```

UltralyticsAssistant commented on ultralytics/data/augment.py:
```diff
@@ -2011,12 +2015,62 @@ def __call__(self, labels: dict[str, Any]) -> dict[str, Any]:
+                        new_segments = []
+                        for mask in new["masks"]:
+                            segs = masks2segments(mask[None])[0]  # Convert single mask
+                            new_segments.append(segs if len(segs) else segments[0])  # Fallback to original if empty
```
⚠️ **HIGH**: If Albumentations returns an empty mask for an instance, this branch reuses `segments[0]`, duplicating the first polygon across multiple objects. That corrupts the annotations and breaks the bbox/segment alignment. We should instead keep the segment list in sync with the transformed objects (e.g., drop the segment when the mask is empty).

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please import `polygons2masks` to avoid a runtime failure, ensure the keypoint reshape uses the transformed object count, and keep transformed segments aligned with their instances instead of reusing the first polygon.

💬 Posted 3 inline comments

onuralpszr committed:
fix: 🐞remove unused import warning for ToTensorV2 in classify_augmentations

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

glenn-jocher committed:
Merge branch 'main' into feat/image-augmentations-pt2

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

Laughing-q committed:
Merge branch 'main' into feat/image-augmentations-pt2

onuralpszr committed:
docs: 📝 simplify compatibility notes for YOLO tasks in data augmentation guide

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

onuralpszr committed:
fix: 🐞 handle keypoint and segment updates in Albumentations transform

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Add the missing `polygons2masks` import so the new segmentation handling path works without a runtime error.

💬 Posted 1 inline comment

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please add the missing `polygons2masks` import to avoid runtime failures, and rework the keypoint reattachment so Albumentations-dropping boxes doesn’t scramble pose labels.

💬 Posted 2 inline comments

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
fix: 🐞 optimize keypoint and segment handling in Albumentations transforms.

onuralpszr committed:
Merge d199a1b242a399e7c143a7f74e7beebcc2486784 into 5e4e0316f49fd318b98f602621033e1d1ebedac2

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

Jserax commented:
i have no errors related to index/shape, but boxes look weird even with NoOp (do nothing): 
1st image - no augmentations
<img width="1920" height="1920" alt="изображение_2025-12-09_172359544" src="https://github.com/user-attachments/assets/602b3d2c-fd63-4403-9f6c-af849e0f039c" />

2nd -  train_cfg["augmentations"] = [
        A.CoarseDropout(num_holes_range=[1, 2], hole_height_range=[0.05, 0.1], hole_width_range=[0.05, 0.1], p=1.0),
    ] 
<img width="1920" height="1920" alt="изображение_2025-12-09_172617516" src="https://github.com/user-attachments/assets/61f19ff4-0d36-485a-b666-6ae98c3e0c84" />

3rd - train_cfg["augmentations"] = [
        A.CoarseDropout(num_holes_range=[1, 2], hole_height_range=[0.7, 0.95], hole_width_range=[0.7, 0.95], p=1.0),
    ]
<img width="1920" height="1920" alt="изображение_2025-12-09_172837132" src="https://github.com/user-attachments/assets/9510642b-f7bf-4abf-ba2f-bd76b8ec4a9e" />
4th - train_cfg["augmentations"] = [
        A.NoOp(p=1.0),
    ]
<img width="1920" height="1920" alt="изображение_2025-12-09_174202463" src="https://github.com/user-attachments/assets/4763aa67-b523-4a5f-86fa-7f1a044d1aed" />

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

glenn-jocher committed:
Merge remote-tracking branch 'origin/main' into feat/image-augmentations-pt2

# Conflicts:
#	docs/mkdocs_github_authors.yaml

glenn-jocher commented:
Hi @onuralpszr — I merged the latest `main` into your branch to clear the conflicts, so it's current again (only `docs/mkdocs_github_authors.yaml` conflicted; I took main's version).

While re-reviewing I think I found the root cause of the "weird boxes" @Jserax reported. Right before the transform the instances are normalized:

```python
labels["instances"].normalize(*im.shape[:2][::-1])  # augment.py:2181
```

which divides bboxes, segments, and keypoints by `(w, h)` into 0–1 space. `A.BboxParams(format="yolo")` expects exactly that normalized space, but `A.KeypointParams(format="xy")` (augment.py:2125) expects **absolute pixel** coordinates — so the keypoints/segment points get passed in as ~0–1 px and are mangled by any spatial op (flip/scale/rotate, and even the coordinate bookkeeping under `NoOp`, since `NoOp`/`CoarseDropout` are in `spatial_transforms`). Denormalizing the points to pixels before the call (× w, h) and normalizing the results back should resolve the desync.

A closely related one: at augment.py:2290 `segment2box(seg, w, h)` returns absolute `xyxy`, and you then set `bboxes = seg_bboxes` and `labels["instances"].update(bboxes=bboxes)` (augment.py:2296 / 2306) while the instances are still normalized `xywh` — so both the format (xyxy vs xywh) and the scale (px vs 0–1) mismatch. The reassigned `.segments`/`.keypoints` are likewise left in px space with `normalized=True`.

Two smaller notes:
- The `pose_kpts[:num_instances]` / `seg_points[:num_instances]` slicing (augment.py:2261 / 2286) assumes Albumentations kept the *first* N instances, but it drops out-of-bounds boxes in place — so keypoints/segments can misalign with their boxes whenever a box is dropped. The `keypoint_labels` you already build could be used to select the exact kept set instead.
- augment.py:2253 `new_kpt_labels[pose_mask]` is a leftover no-op expression you can delete.

Remaining to merge: fix the coordinate-space handling above, add a small test in `tests/` running seg + pose through a spatial Albumentations transform, and re-run CI. Happy to take another look once it's updated!

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into feat/image-augmentations-pt2

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

glenn-jocher committed:
Merge main into feat/image-augmentations-pt2

glenn-jocher committed:
Skip Albumentations sample when distortions drop transformed points

GridDistortion, PiecewiseAffine, ElasticTransform and OpticalDistortion
remap keypoints through an index mask and silently return fewer points
than were passed in. The keypoint/segment reshape then raised ValueError
or mis-assigned points to instances, crashing pose and segment training.
Leave the sample untouched when the returned point count differs.

UltralyticsAssistant committed:
Merge branch 'main' into feat/image-augmentations-pt2

glenn-jocher commented:
Closing this one: keypoints, segments and masks support for Albumentations landed on main via #25633, which routes spatial vs pixel transforms by type and carries everything through a single `A.Compose` with `BboxParams` + `KeypointParams`. The branch now carries no remaining changes. Thanks @onuralpszr for driving this forward — the shipped design benefited from the groundwork here.

glenn-jocher commented:
Reopening — this PR was closed in error during the cleanup sweep. The 2026-08-20 maintenance pass had stripped the branch's remaining content (its classification-pipeline Albumentations support was never delivered by #25633, which covered the detect-path spatial routing). Restoring the pre-08-20 feature content now and deconflicting it with current main.

glenn-jocher commented:
Review findings (the risky parts hold up under direct testing: `A.from_dict` restores trainer-serialized dicts from #25891 including YAML round-trip, and the module-level class pickles cleanly for spawn workers):

**Major — `data/augment.py:2930`:** `AlbumentationsTransform` never calls `set_random_seed()`, so classification training with custom Albumentations silently ignores `seed`/`deterministic` — unlike the detect path, which seeds. Mirror the detect path's seeding.

Minors: missing albumentations raises a bare ImportError in the classify path while the identical input is silently tolerated on the detect path (`:2926`) — pick one behavior for both; every DataLoader worker replays the same Albumentations RNG stream, dividing augmentation diversity by worker count (`:3003`) — fold worker id/seed into the transform seed.
